pkg/parser/expression.go

```diff
@@ -0,0 +1,402 @@
+		}
+		return
+	}
+	if op == OP_AND || op == OP_OR {
+		if leftType != BOOL_TYPE {
+			p.appendErrorForToken("'"+op.String()+"' takes bool type, found "+leftType.Format(), tok)
+		}
+		return
+	}
+}
+
+func (p *Parser) curPrecedence() precedence {
+	if p, ok := precedences[p.cur.Type]; ok {
+		return p
+	}
+	return LOWEST
+}
+
+func (p *Parser) parseLiteral(scope *scope) Node {
+	tok := p.cur
+	tt := tok.TokenType()
+	switch tt {
+	case lexer.STRING_LIT:
+		p.advance()
+		return &StringLiteral{Token: tok, Value: tok.Literal}
+	case lexer.NUM_LIT:
+		p.advance()
+		val, err := strconv.ParseFloat(tok.Literal, 64)
+		if err != nil {
+			p.appendError(err.Error())
+			return nil
+		}
+		return &NumLiteral{Token: tok, Value: val}
+	case lexer.TRUE, lexer.FALSE:
+		p.advance()
+		return &Bool{Token: tok, Value: tt == lexer.TRUE}
+	case lexer.LBRACKET:
+		return p.parseArrayLiteral(scope)
+	case lexer.LCURLY:
+		return p.parseMapLiteral(scope)
+	}
+	return nil
+}
+
+func (p *Parser) parseArrayLiteral(scope *scope) Node {
+	tok := p.cur
+	p.advance() // advance past [
+	elements := p.parseExprList(scope)
+	if elements == nil {
+		return nil // previous error
+	}
+	if !p.assertToken(lexer.RBRACKET) {
+		return nil
+	}
+	p.advance() // advance past ]
+	if len(elements) == 0 {
+		return &ArrayLiteral{Token: tok, T: EMPTY_ARRAY}
+	}
+	types := make([]*Type, len(elements))
+	for i, e := range elements {
+		types[i] = e.Type()
+	}
+	t := &Type{Name: ARRAY, Sub: p.combineTypes(types)}
+	return &ArrayLiteral{Token: tok, Elements: elements, T: t}
+}
+
+func (p *Parser) parseExprList(scope *scope) []Node {
+	list := []Node{}
+	tt := p.cur.TokenType()
+	for !p.isAtEOL() && tt != lexer.RPAREN && tt != lexer.RBRACKET {
+		n := p.parseExpr(scope, LOWEST)
+		if n == nil {
+			return nil // previous error
+		}
+		list = append(list, n)
+		tt = p.cur.TokenType()
+	}
+	return list
+}
+
+func (p *Parser) combineTypes(types []*Type) *Type {
+	combinedT := types[0]
+	for _, t := range types[1:] {
+		if combinedT.Accepts(t) {
+			continue
+		}
+		if t.Accepts(combinedT) {
+			combinedT = t
+			continue
+		}
+		return ANY_TYPE
+	}
+	return combinedT
+}
+
+func (p *Parser) parseMapLiteral(scope *scope) Node {
+	tok := p.cur
+	p.advance() // advance past {
+	pairs, order := p.parseMapPairs(scope)
+	if pairs == nil {
+		return nil // previous error
+	}
+	if !p.assertToken(lexer.RCURLY) {
+		return nil
+	}
+	p.advance() // advance past }
+	if len(pairs) == 0 {
+		return &MapLiteral{Token: tok, T: EMPTY_MAP}
+	}
+	types := make([]*Type, 0, len(pairs))
+	for _, n := range pairs {
+		types = append(types, n.Type())
+	}
+	t := &Type{Name: MAP, Sub: p.combineTypes(types)}
+	return &MapLiteral{Token: tok, Pairs: pairs, Order: order, T: t}
+}
+
+func (p *Parser) parseMapPairs(scope *scope) (map[string]Node, []string) {
+	pairs := map[string]Node{}
+	var order []string
+	tt := p.cur.TokenType()
+
+	for !p.isAtEOL() && tt != lexer.RCURLY {
```

Comment:
i thought a map literal could span multiple lines, in which case it seems wrong to terminate parsing at EOL?

Reply:
yes - I haven't implemented this yet. I think this needs more thought, especially for array literals:  what if they part of an expression?
I'll circle back on this later if that's ok?

Reply:
sure, that's ok. i suspected this was the case.